### ISSUE TITLE
False positive not-callable on type-annotated post-initialized class member

### ISSUE THREAD

**Issue by Strilanc** [PyCQA/pylint#2562]
### Steps to reproduce

Insert the following python code into a file `x.py`

```python
class C:
    X = None  # type: C
    def __call__(self, *args, **kwargs):
        return self
C.X = C()

_ = C.X(5)
```

2. Run pylint with `not-callable` enabled

`pylint --disable=all --enable=not-callable --score=no x.py`

### Current behavior

Pylint outputs an error

```
************* Module x
x.py:8:4: E1102: C.X is not callable (not-callable)
```

This is likely because pylint is looking at the temporary value `None` instead of trusting the type annotation `type: C`. (Note that `mypy` trusts the type annotation.)

### Expected behavior

Consider not outputing an error in this case, in order to make it possible to work around the fact that it is not possible to initialize class variables of the same type as the enclosing class until after the class is initialized, so it is necessary to have some sort of placeholder of an incorrect type.

### pylint --version output

```
pylint 2.1.1
astroid 2.1.0-dev
Python 3.5.3 (default, Jan 19 2017, 14:11:04) 
[GCC 6.3.0 20170118]
```

and separately

```
pylint 2.1.1
astroid 2.0.4
Python 3.5.3 (default, Jan 19 2017, 14:11:04) 
[GCC 6.3.0 20170118]
```

**Comment by Strilanc** [comment]
Note that there is a workaround that works on both mypy and pylint (at least for now): before the class, declare a dummy method that returns the class type (specified as a string), and initialize the fields using the dummy method.

**Comment by PCManticore** [comment]
pylint doesn't interpret type annotations at all, instead it relies on inference to figure out what the actual type of an object is. In this case, it might determine it's an union of `NoneType` and `C` which is why it would emit this error for you. That's definitely a bug, but it's also tripped by the fact that later on you reassign the X member to an actual callable. 
Unfortunately until we'd change pylint to have type inference based from type annotations, there's not much we can do.

**Comment by uliantsev-a** [comment]
@PCManticore, could we check condition of callable before checking to not-callable. For pass of this example:
```
if callable(C.X):
    _ = C.X(5)
```

**Comment by Pierre-Sassoulas** [comment]
The discussion about taking into account type annotations is in #4813. I don't think it would be wise to add small workaround before we resolve that bigger issue. The maintenance cost is just too high.
